azurelinuxagent/common/protocol/extensions_goal_state.py

```diff
@@ -97,29 +98,68 @@ def compare(from_extensions_config, from_vm_settings):
+            if from_vm_settings.host_ga_plugin_version > FlexibleVersion("1.0.8.112"):
+                compare_attributes(first, second, "status_upload_blob")
+                compare_attributes(first, second, "status_upload_blob_type")
+            compare_attributes(first, second, "required_features")
+            compare_attributes(first, second, "on_hold")
+            compare_array(first.agent_manifests, second.agent_manifests, compare_agent_manifests, "agent_manifests")
+            compare_array(first.extensions, second.extensions, compare_extensions, "extensions")
+
+        def compare_agent_manifests(first, second):
+            compare_attributes(first, second, "family")
+            compare_attributes(first, second, "version")
+            compare_attributes(first, second, "uris")
+
+        def compare_extensions(first, second):
+            compare_attributes(first, second, "name")
+            compare_attributes(first, second, "version")
+            compare_attributes(first, second, "state")
+            compare_attributes(first, second, "supports_multi_config")
+            compare_attributes(first, second, "manifest_uris")
+            compare_array(first.settings, second.settings, compare_settings, "settings")
+
+        def compare_settings(first, second):
+            compare_attributes(first, second, "name")
+            compare_attributes(first, second, "sequenceNumber")
+            compare_attributes(first, second, "publicSettings")
+            compare_attributes(first, second, "protectedSettings")
+            compare_attributes(first, second, "certificateThumbprint")
+            compare_attributes(first, second, "dependencyLevel")
+            compare_attributes(first, second, "state")
+
+        def compare_array(first, second, comparer, name):
+            if len(first) != len(second):
+                raise Exception("Number of items in {0} mismatch: {1} != {2}".format(name, len(first), len(second)))
+            for i in range(len(first)):
+                context.append("{0}[{1}]".format(name, i))
+                try:
+                    comparer(first[i], second[i])
+                finally:
+                    context.pop()
+
+        def compare_attributes(first, second, attribute):
+            context.append(attribute)
+            try:
+                first_value = getattr(first, attribute)
+                second_value = getattr(second, attribute)
+
+                if first_value != second_value:
+                    raise Exception("[{0}] != [{1}] (Attribute: {2})".format(first_value, second_value, ".".join(context)))
+            finally:
+                context.pop()
+
+        try:
+            compare_goal_states(from_extensions_config, from_vm_settings)
+        except Exception as exception:
+            raise GoalStateMismatchError("Mismatch in Goal States [Incarnation {0}] != [Etag: {1}]: {2}".format(from_extensions_config.id, from_vm_settings.id, ustr(exception)))
```

Comment:
So this would still raise as soon as it detects the first mismatch right? I can't understand the use of `context` in the new changes, could you add a comment or a sample or something of how the debug messages would look incase of a mismatch?

Reply:
the unit tests have samples (they check the error message). For example, when comparing the state in an instance of extension settings, the context would be something similar to "extensions[0].settings[0].state"

yes, this raises on the first diff. this is intentional.

Reply:
(the full message would be something like "[Incarnation 1] != [Etag: 1234]: [enable] != [bad-state] (Attribute: extensions[0].settings[0].state"

Reply:
> the unit tests have samples (they check the error message). For example, when comparing the state in an instance of extension settings, the context would be something similar to "extensions[0].settings[0].state"

Ahh right, thanks for pointing it out!